src/server.ts

```diff
@@ -120,20 +172,83 @@ app.patch(
   }
 );
 
-app.delete(`/movies/:id`, requestLogger, (req: Request, res: Response) => {
+app.delete('/movies/:id', requestLogger, (req: Request, res: Response) => {
   const id = req.params.id;
   if (movies.find((m) => m.id === id)) {
     movies = movies.filter((m) => m.id !== id);
 
     res.send(`movie with id ${id} deleted`);
   } else {
-    res.statusCode = 404;
-    res.send('Invalid movie ID.');
+    res.status(404).json({ error: 'Invalid movie ID.' });
   }
 });
 
 app.use(errorHandler);
 
+const authRouter = express.Router();
+
+authRouter.post('/registration', (req: Request, res: Response) => {
+  const { name, pass } = req.body;
+  const user = db.find((user) => user.name === name);
+  if (user) {
+    res.status(409).json({ error: `User with name ${name} exists` });
+  } else {
+    db.push({
+      name: name,
+      pass: pass,
+      role: 'user',
+      favMovies: [],
+    });
+
+    res.status(201).json({ message: `User ${name} registered` });
+  }
+});
+
+authRouter.post('/login', (req: Request, res: Response) => {
+  const { name, pass } = req.body;
+  const user = db.find((user) => user.name === name && user.pass == pass);
+  if (!user) {
+    res.status(404).json({ error: 'User or password not found' }).end();
+  } else {
+    const token = encrypt({ name: user.name, role: user.role });
+    res.status(201).json({ token: token }).end();
+  }
+});
+
+const authOnly = (req: Request, res: Response, next: NextFunction) => {
+  if (!req.user) {
+    res.status(404).json({ error: 'User not found' });
+    return;
+  }
+  next();
+};
+
+app.get('/users/:name', authOnly, (req: Request, res: Response) => {
+  if (req.params.name === req.user.name || req.user.role === 'admin') {
+    const user = db.find((user) => user.name === req.params.name);
+    if (user) {
+      res.status(201).json({ user });
+    } else {
+      res.status(404).json({ error: 'User not found' });
+    }
+  } else {
+    res.status(403).json({ error: 'Forbidden' });
+  }
+});
+
+app.patch('/users/:name', authOnly, (req: Request, res: Response) => {
```

Comment:
It's overhead, passing user name as request argument. User can't change other than his own movies, so after authorization you will have required user name.
Your approach isn't wrong, but less efficient.

Reply:
Could you please help me with endpoint name for enhanced approach?

My implementation looks self-explaining - favMovie is part of /users/:id resource which can be updated by PATCH

Reply:
For example '/favourites', and set array of movies ids in the body. After decoding token you will get user name (req.user.name), then just assign passed movies to this user.

Reply:
Anyway, it's nitpick, it's working fine already 👌

Reply:
thank you, I'll update it after 5th Task (DB) is implemented